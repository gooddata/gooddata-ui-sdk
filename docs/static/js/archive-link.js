document.addEventListener('DOMContentLoaded', function () {
    function createArchiveLink() {
        const link = document.createElement('a');
        link.className = 'dropdown-item';
        
        const currentHostname = window.location.hostname;
        if (currentHostname.includes('netlify.app')) {
            // For Netlify previews, link to the versions page within the current preview site
            link.href = window.location.origin + '/docs/gooddata-ui/versions/';
        } else if (currentHostname === 'localhost' || currentHostname === '127.0.0.1') {
            // For local testing, link to the versions page
            link.href = '/versions/';
        } else {
            // For production - link to the unversioned archive page
            link.href = 'https://www.gooddata.com/docs/gooddata-ui/versions/';
        }
        
        // Safe DOM manipulation - XSS protected
        const icon = document.createElement('i');
        icon.className = 'fas fa-archive me-2';
        icon.setAttribute('aria-hidden', 'true');
        
        link.appendChild(icon);
        link.appendChild(document.createTextNode('Archive'));
        
        // Add descriptive text for screen readers
        link.setAttribute('aria-label', 'View all versions in archive');
        
        return link;
    }
    
    // Select the dropdown menu and add the archive link
    const dropdownMenus = document.getElementsByClassName('dropdown-menu');
    for (let i = 0; i < dropdownMenus.length; i++) {
        const archiveLink = createArchiveLink();
        dropdownMenus[i].appendChild(archiveLink);
    }
});