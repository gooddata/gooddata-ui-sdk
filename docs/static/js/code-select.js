(function ($) {
    $(document).ready(function () {
        $(".gd-docs-code-select__tab").on("click", setActive);
        $(".gd-docs-code-select__osdetection .gd-docs-code-select__tab").each(osDetection);
    });

    var setActive = function() {
        if(!$(this).hasClass("active")) {
            $(this).siblings().removeClass("active");
            $(this)
                .addClass("active")
                .parent()
                .parent()
                .find(".gd-docs-code-select__code")
                .removeClass("active")
                .filter(".lang-" + $(this).data("lang"))
                .addClass("active");
        }
    }

    var osDetection = function() {
        if(window.navigator.userAgent.indexOf($(this).data("user-agent")) >= 0) {
            $(this).trigger("click");
        }
    }

}(jQuery));