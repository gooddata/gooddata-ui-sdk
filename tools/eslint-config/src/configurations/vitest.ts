// (C) 2025 GoodData Corporation

import type { IConfiguration } from "../types.js";

const configuration: IConfiguration<"@vitest"> = {
    packages: [
        {
            name: "@vitest/eslint-plugin",
            version: "1.3.5",
        },
    ],
    plugin: "@vitest",
    extends: ["plugin:@vitest/legacy-recommended"],
    rules: {
        "@vitest/expect-expect": "off",
        "@vitest/no-commented-out-tests": "warn",
        "@vitest/valid-title": "warn",
        "@vitest/no-disabled-tests": "warn",
        "@vitest/no-focused-tests": "warn",
        "@vitest/no-identical-title": "warn",
        "@vitest/valid-expect": "warn",
    },
};

export default configuration;
