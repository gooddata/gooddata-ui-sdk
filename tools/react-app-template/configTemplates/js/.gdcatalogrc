{
  "hostname": "https://public-examples.gooddata.com",
  "workspaceId": "demo",
  "output": "src/catalog.ts",
  "backend": "tiger",
  "demo": true
}