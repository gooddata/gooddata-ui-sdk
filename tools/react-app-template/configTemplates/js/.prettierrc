{
    "printWidth": 110,
    "tabWidth": 4,
    "trailingComma": "all"
}
