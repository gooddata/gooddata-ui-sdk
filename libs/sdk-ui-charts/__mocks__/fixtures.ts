// (C) 2007-2019 GoodData Corporation
import { LegacyExecutionRecording, legacyRecordedDataView } from "@gooddata/sdk-backend-mockingbird";

import * as rec from "./recordings/playlist.js";
import { DataViewFacade } from "@gooddata/sdk-ui";

function legacyRecordedDataFacade(rec: LegacyExecutionRecording): DataViewFacade {
    return DataViewFacade.for(legacyRecordedDataView(rec));
}

//
// new fixtures
//

export const testWorkspace = "testWorkspace";

//
// Area chart fixtures
//

export const areaChartWith3MetricsAndViewByAttribute = legacyRecordedDataFacade(
    rec.AreaChartWith3MetricsAndViewByAttribute,
);

export const areaChartWithMeasureViewByAndStackBy = legacyRecordedDataFacade(
    rec.AreaChartWithMeasureViewByAndStackBy,
);

//
// Bar chart fixtures
//

export const barChartWith4MetricsAndViewByTwoAttributes = legacyRecordedDataFacade(
    rec.BarChartWith4MetricsAndViewByTwoAttributes,
);
export const barChartWithStackByAndViewByAttributes = legacyRecordedDataFacade(
    rec.BarChartWithStackByAndViewByAttributes,
);
export const barChartWithViewByAttribute = legacyRecordedDataFacade(rec.BarChartWithViewByAttribute);

export const barChartWith3MetricsAndViewByAttribute = legacyRecordedDataFacade(
    rec.BarChartWith3MetricsAndViewByAttribute,
);
export const barChartWith3MetricsAndViewByAttributeFunformat = legacyRecordedDataFacade(
    rec.BarChartWith3MetricsAndViewByAttributeFunformat,
);

export const barChartWith3MetricsAndViewByAttributePercInFormat = legacyRecordedDataFacade(
    rec.BarChartWith3MetricsAndViewByAttributePercInFormat,
);

export const barChartWithPopMeasureAndViewByAttribute = legacyRecordedDataFacade(
    rec.BarChartWithPopMeasureAndViewByAttribute,
);

export const barChartWithPopMeasureAndViewByAttributeX6 = legacyRecordedDataFacade(
    rec.BarChartWithPopMeasureAndViewByAttributeX6,
);

export const barChartWithPreviousPeriodMeasure = legacyRecordedDataFacade(
    rec.BarChartWithPreviousPeriodMeasure,
);
export const barChartWithSingleMeasureAndNoAttributes = legacyRecordedDataFacade(
    rec.BarChartWithSingleMeasureAndNoAttributes,
);

export const barChartWithPreviousPeriodMeasureX6 = legacyRecordedDataFacade(
    rec.BarChartWithPreviousPeriodMeasureX6,
);

//
// Bubble chart
//
export const bubbleChartWith2MetricsAndAttributeNoPrimaries = legacyRecordedDataFacade(
    rec.BubbleChartWith2MetricsAndAttributeNoPrimaries,
);

export const bubbleChartWith1Metric = legacyRecordedDataFacade(rec.BubbleChartWith1Metric);

export const bubbleChartWith3MetricsAndAttribute = legacyRecordedDataFacade(
    rec.BubbleChartWith3MetricsAndAttribute,
);
export const bubbleChartWith3MetricsAndAttributeNullsInData = legacyRecordedDataFacade(
    rec.BubbleChartWith3MetricsAndAttributeNullsInData,
);

//
// Combo chart
//
export const comboWithTwoMeasuresAndViewByAttribute = legacyRecordedDataFacade(
    rec.ComboChartWithTwoMeasuresViewByAttribute,
);

export const comboChartWithTwoMeasuresViewByAttributeNoBuckets = legacyRecordedDataFacade(
    rec.ComboChartWithTwoMeasuresViewByAttributeNoBuckets,
);

export const comboChartWithTwoMeasuresViewByAttributePercformat = legacyRecordedDataFacade(
    rec.ComboChartWithTwoMeasuresViewByAttributePercformat,
);

//
//
//
export const headlineWithOneMeasure = legacyRecordedDataFacade(rec.HeadlineWithOneMeasure);
export const headlineWithOneMeasureWithIdentifier = legacyRecordedDataFacade(
    rec.HeadlineWithOneMeasureWithIdentifier,
);
export const headlineWithTwoMeasures = legacyRecordedDataFacade(rec.HeadlineWithTwoMeasures);
export const headlineWithTwoMeasuresWithIdentifier = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresWithIdentifier,
);
export const headlineWithTwoMeasuresFirstEmpty = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresFirstEmpty,
);
export const headlineWithTwoMeasuresSecondEmpty = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresSecondEmpty,
);
export const headlineWithTwoMeasuresBothEmpty = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresBothEmpty,
);
export const headlineWithTwoMeasuresBothZero = legacyRecordedDataFacade(rec.HeadlineWithTwoMeasuresBothZero);
export const headlineWithTwoMeasuresFirstZero = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresFirstZero,
);
export const headlineWithTwoMeasuresSecondZero = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresSecondZero,
);
export const headlineWithTwoMeasuresBothSame = legacyRecordedDataFacade(rec.HeadlineWithTwoMeasuresBothSame);

//
// Heatmap recordings
//

export const heatMapWithEmptyCells = legacyRecordedDataFacade(rec.HeatMapWithEmptyCells);
export const heatMapWithMetricRowColumn = legacyRecordedDataFacade(rec.HeatMapWithMetricRowColumn);
//
// Pie chart recordings
//

export const pieChartWithMetricsOnly = legacyRecordedDataFacade(rec.PieChartWithMetricsOnly);
export const pieChartWithMetricsOnlyFundata = legacyRecordedDataFacade(rec.PieChartWithMetricsOnlyFundata);

//
// Scatter plot recording
//

export const scatterPlotWith2MetricsAndAttributeNullsInData = legacyRecordedDataFacade(
    rec.ScatterPlotWith2MetricsAndAttributeNullsInData,
);
export const scatterPlotWith2MetricsAndAttributeWithPrimary = legacyRecordedDataFacade(
    rec.ScatterPlotWith2MetricsAndAttributeWithPrimary,
);

//
// Treemap recordings
//

export const treemapWithMetricAndStackByAttribute = legacyRecordedDataFacade(
    rec.TreemapWithMetricAndStackByAttribute,
);
export const treemapWithMetricAndViewByAttribute = legacyRecordedDataFacade(
    rec.TreemapWithMetricAndViewByAttribute,
);
export const treemapWithTwoMetricsAndStackByAttribute = legacyRecordedDataFacade(
    rec.TreemapWithTwoMetricsAndStackByAttribute,
);
export const treemapWithMetricViewByAndStackByAttribute = legacyRecordedDataFacade(
    rec.TreemapWithMetricViewByAndStackByAttribute,
);
