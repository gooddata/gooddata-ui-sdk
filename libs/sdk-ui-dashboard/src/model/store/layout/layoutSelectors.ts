// (C) 2021-2023 GoodData Corporation
import { createSelector } from "@reduxjs/toolkit";
import {
    ObjRef,
    objRefToString,
    IWidget,
    isKpiWidget,
    isInsightWidget,
    IDashboardLayout,
    IInsightWidget,
    IKpiWidget,
    IDrillToLegacyDashboard,
    InsightDrillDefinition,
    IDrillDownReference,
} from "@gooddata/sdk-model";
import { invariant } from "ts-invariant";
import isEmpty from "lodash/isEmpty.js";

import { DashboardSelector, DashboardState } from "../types.js";
import { ExtendedDashboardWidget, isCustomWidget } from "../../types/layoutTypes.js";
import { UndoableCommand, createUndoableCommandsMapping } from "../_infra/undoEnhancer.js";
import { ObjRefMap, newMapForObjectWithIdentity } from "../../../_staging/metadata/objRefMap.js";
import { selectFilterContextFilters } from "../filterContext/filterContextSelectors.js";
import { filterContextItemsToDashboardFiltersByWidget } from "../../../converters/index.js";
import { createMemoizedSelector } from "../_infra/selectors.js";
import { IDashboardFilter, ILayoutCoordinates } from "../../../types.js";
import {
    isInsightPlaceholderWidget,
    isKpiPlaceholderWidget,
    isPlaceholderWidget,
} from "../../../widgets/index.js";

import { LayoutStash, LayoutState } from "./layoutState.js";
import { isItemWithBaseWidget, getWidgetCoordinates } from "./layoutUtils.js";
import { DashboardLayoutCommands } from "../../commands/index.js";

const selectSelf = createSelector(
    (state: DashboardState) => state,
    (state) => state.layout,
);

/**
 * This selector returns current layout's stash. This stash can contain items that were removed from the layout with the
 * intent of further using the item elsewhere on the layout. The stash is a mapping of stashIdentifier to an array
 * of stashed items. The stash identifiers and stash usage is fully under control of the user.
 *
 * @internal
 */
export const selectStash: DashboardSelector<LayoutStash> = createSelector(
    selectSelf,
    (layoutState: LayoutState) => {
        return layoutState.stash;
    },
);

/**
 * This selector returns commands that impacted the layout and can now be undone.
 *
 * @internal
 */
export const selectUndoableLayoutCommands: DashboardSelector<UndoableCommand<DashboardLayoutCommands>[]> =
    createSelector(selectSelf, (layoutState: LayoutState) => {
        return createUndoableCommandsMapping(layoutState);
    });

/**
 * This selector returns dashboard's layout. It is expected that the selector is called only after the layout state
 * is correctly initialized. Invocations before initialization lead to invariant errors.
 *
 * @public
 */
export const selectLayout: DashboardSelector<IDashboardLayout<ExtendedDashboardWidget>> = createSelector(
    selectSelf,
    (layoutState: LayoutState) => {
        invariant(layoutState.layout, "attempting to access uninitialized layout state");

        return layoutState.layout;
    },
);

/**
 * This selector returns the basic dashboard layout that does not contain any client-side extensions.
 *
 * This selector exists because analytical backend impls are not yet ready to handle persistence of custom
 * widgets (that may have arbitrary payloads). The selector is used only in save and saveAs command handlers,
 * where it obtains the layout without any custom widgets and persists that. Note that the save/saveAs
 * handlers will not wipe the custom widgets from the state during the save - so at this point the custom
 * widgets are treated as client-side extensions.
 *
 * Note: this selector also intentionally removes empty sections; dashboard cannot cope with them and
 * they may readily appear if user adds section full of custom widgets and then does saveAs; such sections
 * would end up empty.
 *
 * @internal
 */
export const selectBasicLayout: DashboardSelector<IDashboardLayout<IWidget>> = createSelector(
    selectLayout,
    (layout) => {
        const dashboardLayout: IDashboardLayout<IWidget> = {
            ...layout,
            sections: layout.sections
                .map((section) => {
                    return {
                        ...section,
                        items: section.items.filter(isItemWithBaseWidget),
                    };
                })
                .filter((section) => !isEmpty(section.items)),
        };

        return dashboardLayout;
    },
);

/**
 * Selects dashboard widgets in an obj ref an array. This map will include both analytical and custom
 * widgets that are placed on the dashboard.
 *
 * @internal
 */
export const selectWidgets: DashboardSelector<ExtendedDashboardWidget[]> = createSelector(
    selectLayout,
    (layout) => {
        const items: ExtendedDashboardWidget[] = [];

        for (const section of layout.sections) {
            for (const item of section.items) {
                if (!item.widget) {
                    continue;
                }

                items.push(item.widget);
            }
        }

        return items;
    },
);

/**
 * Selects dashboard widgets in an obj ref to widget map. This map will include both analytical and custom
 * widgets that are placed on the dashboard.
 *
 * @internal
 */
export const selectWidgetsMap: DashboardSelector<ObjRefMap<ExtendedDashboardWidget>> = createSelector(
    selectWidgets,
    (widgets) => {
        return newMapForObjectWithIdentity(widgets);
    },
);

/**
 * Selects widget by its ref (including custom widgets).
 *
 * @remarks
 * To limit the scope only to analytical widgets, use {@link selectAnalyticalWidgetByRef}.
 *
 * @alpha
 */
export const selectWidgetByRef: (
    ref: ObjRef | undefined,
) => DashboardSelector<ExtendedDashboardWidget | undefined> = createMemoizedSelector(
    (ref: ObjRef | undefined) =>
        createSelector(selectWidgetsMap, (widgetMap): ExtendedDashboardWidget | undefined => {
            if (!ref) {
                return undefined;
            }

            return widgetMap.get(ref);
        }),
);

/**
 * Selects analytical widget by its ref. This selector will return undefined if the provided
 * widget ref is for a custom widget.
 *
 * @remarks
 * To include custom widgets as well, use {@link selectWidgetByRef}.
 *
 * @alpha
 */
export const selectAnalyticalWidgetByRef: (
    ref: ObjRef | undefined,
) => DashboardSelector<IWidget | undefined> = createMemoizedSelector((ref: ObjRef | undefined) =>
    createSelector(selectWidgetsMap, (widgetMap): IWidget | undefined => {
        if (!ref) {
            return undefined;
        }

        const widget = widgetMap.get(ref);

        if (!widget || isCustomWidget(widget)) {
            return undefined;
        }

        return widget;
    }),
);

/**
 * @alpha
 */
export const selectIgnoredDrillDownHierarchiesByWidgetRef: (
    ref: ObjRef,
) => DashboardSelector<IDrillDownReference[]> = createMemoizedSelector((ref: ObjRef) =>
    createSelector(selectAnalyticalWidgetByRef(ref), (widget) => widget?.ignoredDrillDownHierarchies ?? []),
);

/**
 * Selects widget drills by the widget ref.
 *
 * @alpha
 */
export const selectWidgetDrills: (
    ref: ObjRef | undefined,
) => DashboardSelector<IDrillToLegacyDashboard[] | InsightDrillDefinition[]> = createMemoizedSelector(
    (ref: ObjRef | undefined) =>
        createSelector(selectAnalyticalWidgetByRef(ref), (widget) => widget?.drills ?? []),
);

/**
 * Selects all filters from filter context converted to filters specific for a widget specified by a ref.
 *
 * @remarks
 * This does NOT resolve things like ignored filters for a widget, etc.
 *
 * @internal
 */
export const selectAllFiltersForWidgetByRef: (ref: ObjRef) => DashboardSelector<IDashboardFilter[]> =
    createMemoizedSelector((ref: ObjRef) => {
        return createSelector(
            selectWidgetByRef(ref),
            selectFilterContextFilters,
            (widget, dashboardFilters) => {
                invariant(widget, `widget with ref ${objRefToString(ref)} does not exist in the state`);
                return filterContextItemsToDashboardFiltersByWidget(dashboardFilters, widget);
            },
        );
    });

const selectAllWidgets = createSelector(selectWidgetsMap, (widgetMap) => {
    return Array.from(widgetMap.values());
});

/**
 * Selects a boolean indicating if the dashboard is empty.
 *
 * @alpha
 */
export const selectIsLayoutEmpty: DashboardSelector<boolean> = createSelector(
    selectAllWidgets,
    (allWidgets) => {
        return allWidgets.length === 0;
    },
);

/**
 * Selects all KPI widgets in the layout.
 *
 * @alpha
 */
export const selectAllKpiWidgets: DashboardSelector<IKpiWidget[]> = createSelector(
    selectAllWidgets,
    (allWidgets) => {
        return allWidgets.filter(isKpiWidget);
    },
);

/**
 * Selects all insight widgets in the layout.
 *
 * @alpha
 */
export const selectAllInsightWidgets: DashboardSelector<ExtendedDashboardWidget[]> = createSelector(
    selectAllWidgets,
    (allWidgets) => {
        return allWidgets.filter(isInsightWidget);
    },
);

/**
 * Selects all custom widgets in the layout.
 *
 * @alpha
 */
export const selectAllCustomWidgets: DashboardSelector<ExtendedDashboardWidget[]> = createSelector(
    selectAllWidgets,
    (allWidgets) => {
        return allWidgets.filter(isCustomWidget);
    },
);

/**
 * Selects all non-custom widgets in the layout.
 *
 * @alpha
 */
export const selectAllAnalyticalWidgets: DashboardSelector<IWidget[]> = createSelector(
    selectAllWidgets,
    (allWidgets) => {
        return allWidgets.filter((w): w is IKpiWidget | IInsightWidget => !isCustomWidget(w));
    },
);

/**
 * Selects a boolean indicating if the dashboard contains at least one non-custom widget.
 *
 * @alpha
 */
export const selectLayoutHasAnalyticalWidgets: DashboardSelector<boolean> = createSelector(
    selectAllAnalyticalWidgets,
    (allAnalyticalWidgets) => {
        return allAnalyticalWidgets.length > 0;
    },
);

/**
 * Selects layout coordinates for a given widget.
 *
 * @alpha
 */
export const selectWidgetCoordinatesByRef: (ref: ObjRef) => DashboardSelector<ILayoutCoordinates> =
    createMemoizedSelector((ref: ObjRef) => {
        return createSelector(selectLayout, (layout): ILayoutCoordinates => {
            const coords = getWidgetCoordinates(layout, ref);
            invariant(coords, `widget with ref ${objRefToString(ref)} does not exist in the state`);
            return coords;
        });
    });

/**
 * @internal
 */
export const selectWidgetPlaceholder: DashboardSelector<ExtendedDashboardWidget | undefined> = createSelector(
    selectAllCustomWidgets,
    (customWidgets) => {
        return customWidgets.find(isPlaceholderWidget);
    },
);

/**
 * @internal
 */
export const selectWidgetPlaceholderCoordinates: DashboardSelector<ILayoutCoordinates | undefined> =
    createSelector(selectWidgetPlaceholder, selectLayout, (widgetPlaceholder, layout) => {
        return widgetPlaceholder ? getWidgetCoordinates(layout, widgetPlaceholder.ref) : undefined;
    });

/**
 * @internal
 */
export const selectInsightWidgetPlaceholder: DashboardSelector<ExtendedDashboardWidget | undefined> =
    createSelector(selectAllCustomWidgets, (customWidgets) => {
        return customWidgets.find(isInsightPlaceholderWidget);
    });

/**
 * @internal
 */
export const selectInsightWidgetPlaceholderCoordinates: DashboardSelector<ILayoutCoordinates | undefined> =
    createSelector(selectInsightWidgetPlaceholder, selectLayout, (widgetPlaceholder, layout) => {
        return widgetPlaceholder ? getWidgetCoordinates(layout, widgetPlaceholder.ref) : undefined;
    });

/**
 * @internal
 */
export const selectKpiWidgetPlaceholder: DashboardSelector<ExtendedDashboardWidget | undefined> =
    createSelector(selectAllCustomWidgets, (customWidgets) => {
        return customWidgets.find(isKpiPlaceholderWidget);
    });

/**
 * @internal
 */
export const selectKpiWidgetPlaceholderCoordinates: DashboardSelector<ILayoutCoordinates | undefined> =
    createSelector(selectKpiWidgetPlaceholder, selectLayout, (widgetPlaceholder, layout) => {
        return widgetPlaceholder ? getWidgetCoordinates(layout, widgetPlaceholder.ref) : undefined;
    });
