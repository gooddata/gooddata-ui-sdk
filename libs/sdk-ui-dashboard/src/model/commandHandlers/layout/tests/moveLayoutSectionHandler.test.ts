// (C) 2021-2024 GoodData Corporation
import { beforeEach, describe, it, expect } from "vitest";
import { DashboardTester, preloadedTesterFactory } from "../../../tests/DashboardTester.js";
import { TestCorrelation } from "../../../tests/fixtures/Dashboard.fixtures.js";
import { MoveLayoutSection, moveLayoutSection, undoLayoutChanges } from "../../../commands/index.js";
import {
    DashboardCommandFailed,
    DashboardLayoutChanged,
    DashboardLayoutSectionMoved,
} from "../../../events/index.js";
import { selectLayout } from "../../../store/layout/layoutSelectors.js";
import { SimpleDashboardIdentifier } from "../../../tests/fixtures/SimpleDashboard.fixtures.js";
import {
    ComplexDashboardIdentifier,
    ComplexDashboardWithReferences,
} from "../../../tests/fixtures/ComplexDashboard.fixtures.js";

describe("move section command handler", () => {
    describe("for any dashboard", () => {
        let Tester: DashboardTester;

        beforeEach(async () => {
            await preloadedTesterFactory((tester) => {
                Tester = tester;
            }, SimpleDashboardIdentifier);
        });

        it("should fail if bad section index specified", async () => {
            const originalLayout = selectLayout(Tester.state());
            const event: DashboardCommandFailed<MoveLayoutSection> = await Tester.dispatchAndWaitFor(
                moveLayoutSection(originalLayout.sections.length, -1, TestCorrelation),
                "GDC.DASH/EVT.COMMAND.FAILED",
            );

            expect(event.payload.reason).toEqual("USER_ERROR");
            expect(event.correlationId).toEqual(TestCorrelation);
        });
    });

    describe("for dashboard with three sections", () => {
        const [FirstSection, SecondSection, ThirdSection] =
            ComplexDashboardWithReferences.dashboard.layout!.sections;

        let Tester: DashboardTester;

        beforeEach(async () => {
            await preloadedTesterFactory((tester) => {
                Tester = tester;
            }, ComplexDashboardIdentifier);
        });

        it("should move first section to the end of the section list", async () => {
            const event: DashboardLayoutSectionMoved = await Tester.dispatchAndWaitFor(
                moveLayoutSection(0, -1),
                "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_MOVED",
            );

            expect(event.payload.section).toEqual(FirstSection);
            expect(event.payload.toIndex).toEqual(2);

            const layout = selectLayout(Tester.state());
            expect(layout.sections).toEqual([SecondSection, ThirdSection, FirstSection]);
        });

        it("should move first section to the end of the section list using absolute index", async () => {
            const event: DashboardLayoutSectionMoved = await Tester.dispatchAndWaitFor(
                moveLayoutSection(0, 2),
                "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_MOVED",
            );

            expect(event.payload.section).toEqual(FirstSection);
            expect(event.payload.toIndex).toEqual(2);

            const layout = selectLayout(Tester.state());
            expect(layout.sections).toEqual([SecondSection, ThirdSection, FirstSection]);
        });

        it("should move first section to the middle of the section list", async () => {
            const event: DashboardLayoutSectionMoved = await Tester.dispatchAndWaitFor(
                moveLayoutSection(0, 1),
                "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_MOVED",
            );

            expect(event.payload.section).toEqual(FirstSection);
            expect(event.payload.toIndex).toEqual(1);

            const layout = selectLayout(Tester.state());
            expect(layout.sections).toEqual([SecondSection, FirstSection, ThirdSection]);
        });

        it("should move second section to the end of the section list", async () => {
            const event: DashboardLayoutSectionMoved = await Tester.dispatchAndWaitFor(
                moveLayoutSection(1, -1),
                "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_MOVED",
            );

            expect(event.payload.section).toEqual(SecondSection);
            expect(event.payload.toIndex).toEqual(2);

            const layout = selectLayout(Tester.state());
            expect(layout.sections).toEqual([FirstSection, ThirdSection, SecondSection]);
        });

        it("should move second section to the end of the section list using absolute index", async () => {
            const event: DashboardLayoutSectionMoved = await Tester.dispatchAndWaitFor(
                moveLayoutSection(1, 2),
                "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_MOVED",
            );

            expect(event.payload.section).toEqual(SecondSection);
            expect(event.payload.toIndex).toEqual(2);

            const layout = selectLayout(Tester.state());
            expect(layout.sections).toEqual([FirstSection, ThirdSection, SecondSection]);
        });

        it("should move second section to the beginning of the section list", async () => {
            const event: DashboardLayoutSectionMoved = await Tester.dispatchAndWaitFor(
                moveLayoutSection(1, 0),
                "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_MOVED",
            );

            expect(event.payload.section).toEqual(SecondSection);
            expect(event.payload.toIndex).toEqual(0);

            const layout = selectLayout(Tester.state());
            expect(layout.sections).toEqual([SecondSection, FirstSection, ThirdSection]);
        });

        it("should move third section to the middle of the section list", async () => {
            const event: DashboardLayoutSectionMoved = await Tester.dispatchAndWaitFor(
                moveLayoutSection(2, 1),
                "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_MOVED",
            );

            expect(event.payload.section).toEqual(ThirdSection);
            expect(event.payload.toIndex).toEqual(1);

            const layout = selectLayout(Tester.state());
            expect(layout.sections).toEqual([FirstSection, ThirdSection, SecondSection]);
        });

        it("should be undoable", async () => {
            // move first section into the middle
            await Tester.dispatchAndWaitFor(
                moveLayoutSection(0, 1),
                "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_MOVED",
            );

            // then move it from middle to the end of the list
            await Tester.dispatchAndWaitFor(
                moveLayoutSection(1, 2),
                "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_MOVED",
            );

            const lastMoveUndone: DashboardLayoutChanged = await Tester.dispatchAndWaitFor(
                undoLayoutChanges(),
                "GDC.DASH/EVT.FLUID_LAYOUT.LAYOUT_CHANGED",
            );

            expect(lastMoveUndone.payload.layout.sections).toEqual([
                SecondSection,
                FirstSection,
                ThirdSection,
            ]);
            expect(selectLayout(Tester.state()).sections).toEqual([
                SecondSection,
                FirstSection,
                ThirdSection,
            ]);

            const firstMoveUndone: DashboardLayoutChanged = await Tester.dispatchAndWaitFor(
                undoLayoutChanges(),
                "GDC.DASH/EVT.FLUID_LAYOUT.LAYOUT_CHANGED",
            );

            expect(firstMoveUndone.payload.layout.sections).toEqual([
                FirstSection,
                SecondSection,
                ThirdSection,
            ]);
            expect(selectLayout(Tester.state()).sections).toEqual([
                FirstSection,
                SecondSection,
                ThirdSection,
            ]);
        });

        it("should emit correct event", async () => {
            await Tester.dispatchAndWaitFor(
                moveLayoutSection(0, 1, TestCorrelation),
                "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_MOVED",
            );

            expect(Tester.emittedEventsDigest()).toMatchSnapshot();
        });
    });
});
