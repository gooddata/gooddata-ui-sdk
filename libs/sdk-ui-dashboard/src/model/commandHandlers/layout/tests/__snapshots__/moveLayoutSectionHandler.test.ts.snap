// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`move section command handler > for dashboard with three sections > should emit correct event 1`] = `
[
  {
    "commandType": "GDC.DASH/CMD.FLUID_LAYOUT.MOVE_SECTION",
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.COMMAND.STARTED",
  },
  {
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_MOVED",
  },
]
`;
